dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackerAccessManager.java

```diff
@@ -59,31 +63,83 @@ public class DefaultTrackerAccessManager implements TrackerAccessManager {
 
   private final AclService aclService;
   private final TrackerOwnershipManager ownershipAccessManager;
-
+  private final ProgramService programService;
+
+  /**
+   * Check the data read permissions and ownership of a tracked entity given the programs for which
+   * the user has metadata access to.
+   *
+   * @return No errors if a user has access to at least one program
+   */
   @Override
   public List<String> canRead(UserDetails user, TrackedEntity trackedEntity) {
     // always allow if user == null (internal process) or user is superuser
     if (user == null || user.isSuper() || trackedEntity == null) {
       return List.of();
     }
 
-    OrganisationUnit ou = trackedEntity.getOrganisationUnit();
-    List<String> errors = new ArrayList<>();
-    // ou should never be null, but needs to be checked for legacy reasons
-    if (ou != null && !user.isInUserSearchHierarchy(ou.getPath())) {
-      errors.add(NO_READ_ACCESS_TO_ORG_UNIT + ": " + ou.getUid());
+    return new ArrayList<>(canRead(user, trackedEntity, programService.getAllPrograms()));
+  }
+
+  private Set<String> canRead(
+      UserDetails user, TrackedEntity trackedEntity, List<Program> programs) {
+
+    if (null == trackedEntity) {
+      return Set.of();
     }
 
     TrackedEntityType trackedEntityType = trackedEntity.getTrackedEntityType();
 
     if (!aclService.canDataRead(user, trackedEntityType)) {
-      errors.add(
+      return Set.of(
           "User has no data read access to tracked entity type: " + trackedEntityType.getUid());
     }
 
+    initializeTrackedEntityOrgUnitParents(trackedEntity);
+
+    Set<String> errors = new HashSet<>();
+
+    List<Program> tetPrograms =
+        programs.stream()
+            .filter(
+                p -> Objects.equals(p.getTrackedEntityType(), trackedEntity.getTrackedEntityType()))
+            .toList();
+
+    if (tetPrograms.isEmpty()) {
+      return Set.of("User has no access to any program");
```

Comment:
@lucaCambi77 I changed this to the error message above, I think it makes it clearer for the user to understand what's going on. Anyway, we are not using this error messages anywhere, we just check whether this method returns values in the collection or not, and then we throw a different error message.